### PR TITLE
Review for library and test files

### PR DESCRIPTION
**Pull request by liuhanyao98**

### CODE REVIEW

File: modeling_tool.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+#ifndef MODELING_TOOL_H
+#define MODELING_TOOL_H
+
+#include <chrono>
+#include <iostream>
+#include <stdexcept>
+
+#include "google/cloud/spanner/client.h"
+
+namespace modelingtool{
```
Comment (zhiwen-fang) [inline modeling_tool.h:23]: could you use modeling_tool?

File: modeling_tool.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+namespace spanner = ::google::cloud::spanner;
+using google::cloud::StatusOr;
+
+const std::int64_t DAYINTERVAL = 60; 
+const std::string TABLE = "TestModels";
+
+int batchUpdateData(spanner::Client readClient, spanner::Client writeClient,
```
Comment (zhiwen-fang) [inline modeling_tool.h:30]: can we use const ReadClient&?

File: modeling_tool.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+           std::int64_t batchSize)  {
+  auto rows = readClient.Read(TABLE, spanner::KeySet::All(),
+  		{"CdsId", "ExpirationTime", "TrainingTime"});
+  int updatedRecord = 0; 
+  spanner::Mutations mutations;
+  std::int64_t i = 0;
+  for(const auto& row : rows) {
+    if(!row) throw std::runtime_error(row.status().message());
```
Comment (zhiwen-fang) [inline modeling_tool.h:38]: it is better to return a error status instead of throwing exception.

File: modeling_tool.h [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    
+    spanner::Value cds = (*row).get(0).value();
```
Comment (zhiwen-fang) [inline modeling_tool.h:40]: how about row->get(0)?

File: modeling_tool.h [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    spanner::Value expiration = (*row).get(1).value();
+    spanner::Value training = (*row).get(2).value();
+
+    std::int64_t cdsId = cds.get<std::int64_t>().value();
+    StatusOr<spanner::Timestamp> expirationTime = expiration.get<spanner::Timestamp>();
+    StatusOr<spanner::Timestamp> trainingTime = training.get<spanner::Timestamp>();
+    if(!trainingTime) {
+      throw std::runtime_error("TrainingTime shouldn't be null.");
```
Comment (zhiwen-fang) [inline modeling_tool.h:48]: ditto

File: modeling_tool.h [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        throw std::runtime_error("Time gap for " + std::to_string(cdsId) + " is not correct.");
+      }
+    }
+    else {
+      spanner::sys_time<std::chrono::nanoseconds> expirationNS = 
+        (*trainingTime).get<spanner::sys_time<std::chrono::nanoseconds>>().value()
+        + DAYINTERVAL*std::chrono::hours(24);
+      spanner::Timestamp newExpiration = spanner::MakeTimestamp(expirationNS).value();
+      mutations.push_back(spanner::UpdateMutationBuilder(
+		  TABLE, {"CdsId", "ExpirationTime", "TrainingTime"})
```
Comment (zhiwen-fang) [inline modeling_tool.h:65]: maybe define some  const char[] for these column names ?

File: modeling_tool.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+		  .EmplaceRow(cdsId, newExpiration, *trainingTime)
+		  .Build());
+      ++i;
+      if(i%batchSize == 0) {
+          writeClient.Commit(mutations);
+    	  auto commit_result = writeClient.Commit(mutations);
```
Comment (zhiwen-fang) [inline modeling_tool.h:71]: const auto&?

File: modeling_tool.h [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    if (!commit_result) {
+     	throw std::runtime_error(commit_result.status().message());
+   	}
+    updatedRecord += mutations.size();
+  }
+  return updatedRecord;
+}
+
+void batchInsertData(google::cloud::spanner::Client client, std::int64_t batchSize) {
+  namespace spanner = ::google::cloud::spanner;
```
Comment (zhiwen-fang) [inline modeling_tool.h:92]: it is not needed because you have these statements out of this method?

File: modeling_tool.h [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    else {
+      spanner::sys_time<std::chrono::nanoseconds> expirationNS = 
+        (*trainingTime).get<spanner::sys_time<std::chrono::nanoseconds>>().value()
+        + DAYINTERVAL*std::chrono::hours(24);
+      spanner::Timestamp newExpiration = spanner::MakeTimestamp(expirationNS).value();
+      mutations.push_back(spanner::UpdateMutationBuilder(
+		  TABLE, {"CdsId", "ExpirationTime", "TrainingTime"})
+		  .EmplaceRow(cdsId, newExpiration, *trainingTime)
+		  .Build());
+      ++i;
```
Comment (zhiwen-fang) [inline modeling_tool.h:68]: Is i == mutation.size() for sure?

File: modeling_tool_test.cc [hunk 10]
```diff
@@ -0,0 +1,224 @@
+#include "modeling_tool.h"
+
+#include <gmock/gmock.h>
+#include "google/cloud/spanner/client.h"
+#include "google/cloud/spanner/mocks/mock_spanner_connection.h"
+#include <google/protobuf/text_format.h>
+
+namespace {
```
Comment (zhiwen-fang) [inline modeling_tool_test.cc:8]: generally, you can put the same the same namespace as implementation.

File: modeling_tool_test.cc [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+using ::testing::_;
+using ::testing::Return;
+using ::testing::Field;
+using google::cloud::StatusOr;
```
Comment (zhiwen-fang) [inline modeling_tool_test.cc:12]: for using statement, it is better to unify the format. i.e. using ::google::cloud::StatusOr;

File: modeling_tool_test.cc [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+        spanner::Client readClient(readConn);
+        spanner::Client writeClient(writeConn);
+        modelingtool::batchUpdateData(readClient, writeClient, 1);
+        FAIL() << "Expected std::runtime_error";
+    }
+    catch(std::runtime_error const & err) {
+        EXPECT_EQ(err.what(), std::string("Time gap for 1 is not correct."));
+    }
+    catch(...) {
+      FAIL() << "Expected std::runtime_error";
```
Comment (zhiwen-fang) [inline modeling_tool_test.cc:186]: Just check the returned status?

File: modeling_tool.h (hunk 9 above)
Comment (liuhanyao98) [reply modeling_tool.h:68]: yes, i += 1 when one more mutation is pushed back to mutations vector for commit.

File: modeling_tool.h (hunk 7 above)
Comment (liuhanyao98) [reply modeling_tool.h:71]: Done

File: modeling_tool.h (hunk 2 above)
Comment (liuhanyao98) [reply modeling_tool.h:30]: we cannot use const since the methods of spanner::client are not const

File: modeling_tool.h (hunk 3 above)
Comment (liuhanyao98) [reply modeling_tool.h:38]: Done, changed to return google::cloud::StatusOr<T>.

File: modeling_tool.h (hunk 6 above)
Comment (liuhanyao98) [reply modeling_tool.h:65]: Done. Use const char*[] for column names now.

File: modeling_tool_test.cc (hunk 12 above)
Comment (liuhanyao98) [reply modeling_tool_test.cc:186]: Done

**Pull request merged**